lib/write-set-tokens.ts

```diff
@@ -0,0 +1,81 @@
+import { types, Set, Range, Char, SetTokens } from './types';
+import * as sets from './sets-lookup';
+
+/**
+ * Takes character code and returns character to be displayed in a set
+ * @param {number} charCode Character code of set element
+ * @returns {string} The string for the sets character
+ */
+export function setChar(charCode: number): string {
+  // We only need to negate first internal '^' if token.not is false
+  return charCode === 94 ? '\\^' :
+    charCode === 92 ? '\\\\' :
+      charCode === 93 ? '\\]' :
+        charCode === 45 ? '\\-' :
+          String.fromCharCode(charCode);
+}
+
+/**
+ * Test if a character set matches a 'set-lookup'
+ * @param {SetTokens} set The set to be tested
+ * @param {Record<string | number, boolean>} map The predefined 'set-lookup'
+ * @returns {boolean} True if the character set corresponds to the 'set-lookup'
+ */
+function isSameSet(set: SetTokens, map: Record<string | number, boolean>): boolean {
+  for (const elem of set) {
+    if (elem.type === types.SET) {
+      return false;
+    }
+    const key = elem.type === types.CHAR ? elem.value : `${elem.from}-${elem.to}`;
+    if (map[key]) {
+      map[key] = false;
+    } else {
+      return false;
+    }
+  }
+  return true;
+}
+
+/**
+ * Writes the tokens for a set
+ * @param {Set} set The set to display
+ * @param {boolean} isNested Whether the token is nested inside another set token
+ * @returns {string} The tokens for the set
+ */
+export function writeSetTokens(set: Set, isNested = false): string {
+  const len = set.set.length;
+  if (len === 1 && isSameSet(set.set, sets.INTS())) {
+    return set.not ? '\\D' : '\\d';
+  } else if (len === 4) {
+    if (isSameSet(set.set, sets.WORDS())) {
+      return set.not ? '\\W' : '\\w';
+    }
+    // Notanychar is only relevant when not nested inside another set token
+    if (set.not && isSameSet(set.set, sets.NOTANYCHAR())) {
```

Comment:
check `len` here too?

Reply:
IThe `len === 4` check in line 49 covers that :)